app/controllers/concerns/two_factor_authenticatable_methods.rb

```diff
@@ -17,7 +17,11 @@ def handle_valid_verification_for_authentication_context(auth_method:)
     if IdentityConfig.store.feature_new_device_alert_aggregation_enabled &&
        user_session[:new_device] != false
       if current_user.sign_in_new_device_at.blank?
-        current_user.update(sign_in_new_device_at: disavowal_event.created_at)
+        current_user.update(
+          sign_in_new_device_at: current_user.events.where(
+            event_type: 'sign_in_notification_timeframe_expired',
+          ).last.created_at,
+        )
```

Comment:
This is an interesting approach. I guess the idea is that the only circumstance where this should be blank is if someone waited long enough for the sign-in notification window to elapse, but then proceeded to finish signing in during the same session?

I think that could work, though would have a few suggestions:

- We should have some sort of fallback in case that event doesn't exist, otherwise this code will currently throw. I might suggest falling back to some other value, and maybe logging, since we don't expect that it should ever happen and would want to know about it?
- This sort of query doesn't have any innate ordering, so "last" doesn't mean how we think it is in how we're using it here. On a related note, I seem to recall events querying being performance-sensitive, so we might try to ensure that some limits are built-in to the query. I'd recommend looking at [`User#second_last_signed_in_at`](https://github.com/18F/identity-idp/blob/c794786f75cb5ada40e237d71e9054e4ea67dbf7/app/models/user.rb#L447-L450) as a good example of ordering and limiting in this way.

Reply:
That seemed to be pretty effective. It's getting some event overflow. But seems to be by changing this:
https://github.com/18F/identity-idp/blob/c794786f75cb5ada40e237d71e9054e4ea67dbf7/app/services/user_alerts/alert_user_about_new_device.rb#L65-L67
        user.sign_in_new_device_at,
        (IdentityConfig.store.new_device_alert_delay_in_minutes +1).minutes.ago,
      ].max
      
 It doesn't get stale events. I need to check it doesn't break the test for that and also add testing you requested above.